moo.js

```diff
@@ -437,66 +433,53 @@
   }
 
   Lexer.prototype.next = function() {
-    if (this.queuedToken) {
-      var queuedToken = this.queuedToken, queuedThrow = this.queuedThrow
-      this.queuedToken = null
-      this.queuedThrow = false
-      if (queuedThrow) {
-        throw new Error(this.formatError(queuedToken, "invalid syntax"))
-      }
-      return queuedToken
+    var index = this.index
+
+    // If a fallback token matched, we don't need to re-run the RegExp
+    if (this.queuedGroup) {
+      var token = this._token(this.queuedGroup, this.queuedText, index)
```

Comment:
If JS had tuples / multi-return, I'd have written all this code the other way round, e.g. `var [group, text] = this._match(index)`.